### ISSUE TITLE
Big radiator and temperature tends to overshoot

### ISSUE THREAD

**Issue by Anycubic** [alexmohr/open-heat#12]
I'm testing a big radiator (1x2 meters), temperature tends to overshoot (i.e. set temperature was 21 and 22 reached + radiator was still pretty hot). I guess that maybe it's a matter of latency, which value can I try to play with (in RadiatorValve.cpp, right?).

**Comment by Anycubic** [comment]
I made some tests, my big radiator was still overshooting a lot. Looking for ideas I saw ISTAtrol code, where `MOT_OPEN_TIME` is double than M`OT_CLOSE_TIME` while in your code is quite the opposite.
So I changed to closeTime values accordingly:

```
if (predictDiff <= 2) {
        closeTime = 5000;
      } else if (predictDiff <= 1.5) {
        closeTime = 4000;
      } else if (predictDiff <= 1) {
        closeTime = 2500;
      } else if (predictDiff <= 0.5) {
        closeTime = 1500;
```

As per attached graph the result is awesome! I'll keep on testing also with my smaller radiators
![2021-11-06_21-59-05](https://user-images.githubusercontent.com/31375394/140627013-af64e651-9d8e-4ae6-8bdf-017f6f39bcf4.jpg)
.

**Comment by alexmohr** [comment]
Awesome, I'll test with these values too

**Comment by Anycubic** [comment]
There must be something wrong in the code. After approx 90 minutes that the radiator was switched on the temperature started to raise and raise. I switched off but the valve didn't close. I had to reset the ESP and immediately the valve closed but it was pretty open. It's like it was behaving at opposite of it was supposed to do, opening instead of closing an then just didn't close at all
![2021-11-07_20-25-15](https://user-images.githubusercontent.com/31375394/140659753-7075410d-dc88-4a19-8292-20f7bf997351.jpg)
![2021-11-07_20-46-23](https://user-images.githubusercontent.com/31375394/140659758-08e08a3d-53ab-4be6-9e98-3684a0a7dda6.jpg)

.

**Comment by alexmohr** [comment]
Make sure the wiring is according to the schematic. I fixed some inconsistentencies a while ago. Motor ground should be pin 12 and vin should be 14. 
I'll test it further tomorrow but its working fine so far for me

**Comment by Anycubic** [comment]
Yes I was aware at a certain point Vin and ground were swapped, that's not the case. Wires are fine, there must be something tricky which makes the valve get crazy, as I said reset and voilà it full closed. I'll try tounderstand why. At a certain point I restarted HA but I don't think this was the issue

**Comment by alexmohr** [comment]
Actually I had the same issue over night. I'll try to fix it

**Comment by Anycubic** [comment]
Good you could reproduce it. It's like if when the set temperature is close to the effective one something starts to go wrong. I wonder also if that "fail to close the valve on shut off" was due to the same bug

**Comment by alexmohr** [comment]
Well I increased the time again to 90s and it's working for the moment. But I'm pretty sure it will break again sooner or later. 
Reading the valve state back is not trivial and needs additional hardware, 7 resistors, 3 OPs and 2 diodes to realize at schmitt trigger based approach as I don't think the approach with another ADC is viable. I'm not sure yet if I'm going to implement this. For the moment the current solution is "good enough"

**Comment by Anycubic** [comment]
Of course being able to know when the valve is actually closed (or opened) is the best thing..So what does happen? The valve position goes out of synch with calculations?
Out of curiosity: isn't going to harm the motor that the valve being closed and  motor still trying to push? Real closing time is around 40 seconds with my valves

**Comment by alexmohr** [comment]
I'm still not sure what happens I probably have to write tests at some point to find bugs like these in a controlled environment. 
The calculations _should_ be done is such way that they cannot run out of sync. 

It's probably not too good for the motor. I can't really tell how this affects long longevity of it. Maybe it reduces the lifespan of the motor a bit. 

I just ordered the necessary parts to implement a "stop detection" on a breadboard at some point. 
This also would make the voltage regulator you asked a couple of days ago mandatory because the circuit would depend on a constant Vcc. But I don't think I'll get around to try this any time soon.

**Comment by Anycubic** [comment]
What about issuing e a "off" command followed by a simulated ESP off/on sequence and "heet".command every on hour or so? I mean if the valve was stuck/out of synch this will restart it properly. Don't know how the swtich on/off could be done properly via software, I mean emulating the unplug/plug which at least closes the valve eventually if it got stuck.

**Comment by alexmohr** [comment]
No I won't implement something like this because it's bad for the battery life (and it's already bad). Either I'll fix the time based approach or the "stop detection" will fix this. The stop detection also should be using less battery.

But to be honest if I'm not getting it a battery lifetime of at least 30 days I'll drop this project anyways.
I have some things in the pipeline to improve battery lifetime like the extended modem sleep over night you suggested.
At the moment it's using about 5% battery in 24h (with heating enabled).

**Comment by Anycubic** [comment]
Which battery are tou talking about? 2 18650 in parallel? I can't believe 30 days only with such a batteries! Btw 15minutes wifi + very deep sleep during night time should be very helpful. As far as I'm concerned if you fix the stuck valve problem your work is pretty good and I'm going to use it in all of my radiators (approx 14...)

**Comment by alexmohr** [comment]
` 2x18650` but probably my batteries are not the best. 
I already implemented the option to set the modem sleep time via mqtt. Automating the through home assitant or a cron job must be done manually. I'll commit it as soon as it's tested. 
I haven't seen the stuck valve in a while so maybe you can test it again?

**Comment by Anycubic** [comment]
Today I will flash the latest available deep sleep release

**Comment by Anycubic** [comment]
I come to the conclusion that your default `openTime`, `closeTime` and `checkIntervalMillis_` are too much aggressive considering my big radiator latency: before the valve could measure ANY change in temperature `VALVE_FULL_ROTATE_TIME` value will be exceeded quite easily (unless you set in many minutes) -> the valve will got out synch/tange. So I'm testing with much lower values and 5 minutes `checkIntervalMillis_`, I'll let you know how it goes.
I guess that the only real solution is being able to know when the valve is actually opened and closed

**Comment by alexmohr** [comment]
I don't think that is the issue because if the valve is opened for 90s the pin which pushes against the actual valve of the radiator will be fully retracted. The latency of your radiator does not matter at all.
But feel free to experiment with other values. 

As I already mentioned before I won't get around to implement the "stop detection" any time soon. The necessary operational amplifiers are scheduled for delivery in January 

Fyi I pushed a change which allows you to configure the modem sleep time via mqtt but It's not exposed on the webinterface.
Check the readme for details.

**Comment by Anycubic** [comment]
ok thank you!

**Comment by Anycubic** [comment]
> Fyi I pushed a change which allows you to configure the modem sleep time via mqtt but It's not exposed on the webinterface. Check the readme for details.

I couldn't find anything about that in the readme, am I still sleeping? :-D

**Comment by alexmohr** [comment]
No you are not. It messed up merging the readme. I'll update it later today. 
`modemsleep/set` and `modemsleep/get` are the mqtt topics. Time is in milliseconds

**Comment by alexmohr** [comment]
But I have to say, that the modemsleep has no significant influence on battery life time. It still drops ~0.4% per hour.

**Comment by Anycubic** [comment]
> But I have to say, that the modemsleep has no significant influence on battery life time. It still drops ~0.4% per hour.

So do you think that is actually moving the motor which is very power hungry? Maybe moving the motor the same amount you are doing is not good for batteries. I also noticed that %opening of the valve is not linear, i.e. opening it 1 second seems a lot compared to say 10 seconds (is it a matter of fluid pressure maybe?). I'm now working with these values, till now got NO crazy valve syndrome (still using `VALVE_FULL_ROTATE_TIME` = 60) but I need to test more:

```
unsigned short openTime = 200U;
unsigned short closeTime = 400U;
```

```
openTime = 1000;
      } else if (predictDiff >= 2) {
        openTime = 1000;
      } else if (predictDiff >= 1.5) {
        openTime = 500;
      } else if (predictDiff >= 1) {
        openTime = 500;
      } else if (predictDiff >= 0.5) {
        openTime = 500;

  closeTime = 1500;
      } else if (predictDiff <= 1.5) {
        closeTime = 800;
      } else if (predictDiff <= 1) {
        closeTime = 800;
      } else if (predictDiff <= 0.5) {
        closeTime = 800;
```

**Comment by alexmohr** [comment]
No it's the esp itself. Heating was turned off over night. The motor is even more power hungry. But the discharge curve of a battery is not linear anyways. I can make definite statements when I'm not updating the devices on a regular basis because that always uses a lot of power.
I'll try out the values you've posted and see if they are working for me too. If they are you can either open a MR or I'll just commit it

**Comment by Anycubic** [comment]
Well it will be interesting to see how they work on a "normal" size radiator, btw need to test those values for more days, here is still not very cold so I can't test properly

**Comment by Anycubic** [comment]
Unfortunately the not closing bug it's still there. I will increase slowly  VALVE_FULL_ROTATE_TIME value to try to find a sweet spot. I will also plug a full time serial monitor trying to understand when the error start ro build up.
Please also note that those my closing numbers are too weak, I'll try with 2000 and 1000 just for now

**Comment by alexmohr** [comment]
The serial monitor is a great idea. Thanks

**Comment by Anycubic** [comment]
> The serial monitor is a great idea. Thanks
In debug mode the BME280 sensor is pretty overheated, it gives more than 1C then usual reading. Are you able to.make a special debug build which is behaving exactly like the deepsleep one but of course without going to sleep? I'd like to debug something which is as similar to reality as possible

**Comment by alexmohr** [comment]
replace `ESP.deepSleep` in https://github.com/alexmohr/open-heat/blob/dev/src/RTCMemory.cpp#L245 with `delay`

**Comment by Anycubic** [comment]
I actually thought the same thing but couldn't believe it was going to be so easy 🤣
But will serial monitor/debugging messages work without setting debug = true via MQTT?

**Comment by alexmohr** [comment]
yes debug mode is just used to disable sleep via mqtt i.e. to enable web interface

**Comment by Anycubic** [comment]
Ok I will test it in the next few days

**Comment by alexmohr** [comment]
I found a possible root cause.
```
Index: src/heating/RadiatorValve.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/heating/RadiatorValve.cpp b/src/heating/RadiatorValve.cpp
--- a/src/heating/RadiatorValve.cpp	(revision d545fadc503d06a291dc5296d872434d0d487b4b)
+++ b/src/heating/RadiatorValve.cpp	(date 1636707779581)
@@ -202,8 +202,11 @@
   }
 
   if (rtc::read().currentRotateTime < 0) {
-    const auto remainingTime
+    auto remainingTime
       = VALVE_FULL_ROTATE_TIME - std::abs(rtc::read().currentRotateTime);
+    if (remainingTime < 0) {
+      remainingTime = 0;
+    }
     rotateTime = std::min(rotateTime, static_cast<unsigned int>(remainingTime));
   }
 
@@ -237,8 +240,12 @@
   }
 
   if (rtc::read().currentRotateTime > 0) {
-    const auto remainingTime
+    auto remainingTime
       = VALVE_FULL_ROTATE_TIME - std::abs(rtc::read().currentRotateTime);
+    if (remainingTime < 0) {
+      remainingTime = 0;
+    }
+
     rotateTime = std::min(rotateTime, static_cast<unsigned int>(remainingTime));
   }
 
```

**Comment by Anycubic** [comment]
I'll try to understand later. Actually I'm logging the valve as we speak

**Comment by Anycubic** [comment]
Wait a minute, one says "<" and the other ">" 🤣

**Comment by alexmohr** [comment]
Because one is close and the other is open.

**Comment by Anycubic** [comment]
I see! I thought that during a merge you mistakenly inverted the sign. So I guess I'll think about all of the night long! :-D
BTW, in my logs today I saw a pretty strange thing:

```
[1;37m[DEBUG] Updating RTCMemory
[1;37m[DEBUG] Updating RTCMemory
[1;32m[INFO]  DEBUG: 0
[1;37m[DEBUG] Sleep times: valveSleep: 299999, mqttSleep: 295521
[1;37m[DEBUG] Updating RTCMemory
[1;37m[DEBUG] Updating RTCMemory
[1;37m[DEBUG] Closing valve for 0ms, currentRotateTime: **-4464ms**, vin: 14, ground: 12
[1;37m[DEBUG] Updating RTCMemory
[1;37m[DEBUG] Updating RTCMemory
[1;32m[INFO]  DEBUG: 0
[1;37m[DEBUG] Sleep times: valveSleep: 299995, mqttSleep: 299978
 Updating RTCMemory
[1;37m[DEBUG] _Opening valve_ for 1000ms, currentRotateTime: **-2464ms**, vin: 14, ground: 12
[1;37m[DEBUG] Updating RTCMemory
[1;32m[INFO]  DEBUG: 0
1;37m[DEBUG] Updating RTCMemory
[1;37m[DEBUG] Updating RTCMemory
[1;37m[DEBUG] _Closing valve_ for 4464ms, currentRotateTime: -4000ms, vin: 14, ground: 12
[1;37m[DEBUG] Updating RTCMemory
[1;37m[DEBUG] Updating RTCMemory
[1;32m[INFO]  DEBUG: 0
Updating RTCMemory
[1;37m[DEBUG] Updating RTCMemory
[1;37m[DEBUG] Closing valve for 464ms, currentRotateTime: -4464ms, vin: 14, ground: 12
[1;37m[DEBUG] Updating RTCMemory
[1;37m[DEBUG] Updating RTCMemory
```
Those where taken from change to "heat" and to "off". There is that initial `currentRotateTime: -4464ms` which I can't understand where is coming from. The log was taken right after flashing the firmware. The first 4 1000ms commands opening where never performed because of that -4464ms and the fifth was done partially which isn't right at all. Also on closing the negative term appears again. From where is this fixed negative offset coming from?

**Comment by Anycubic** [comment]
Uhmmm is your fix going to avoid those negative terms?  Just guessing here.

```
if (remainingTime < 0) {
+      remainingTime = 0;
```

**Comment by alexmohr** [comment]
Negative rotate times are on purpose. 0 would be half way between fully open and fully closed. Rotate time of = max rotate time = fully open. Negative full rotate time is fully closed. 
The patch I posted should prevent that the subtraction wizh remaining time yields strange results when casting into and unsigned int. Although as it selects the minimum and the value is limit in the RTC storage it should not matter much.

About your trace: 
I guess you've updated the device via OTA? In this case the data from ram is kept  the restart. Only a power cycle with unplugging the esp will erase them. 
Also please post the non truncated log instead of snippets. 
Changing the mode to off should also close the valve immediately and fully so something does not add up

**Comment by alexmohr** [comment]
If the current rotate time is negative no range check if performed either ... https://github.com/alexmohr/open-heat/blob/dev/src/heating/RadiatorValve.cpp#L246

**Comment by Anycubic** [comment]
> 
> About your trace: I guess you've updated the device via OTA? In this case the data from ram is kept the restart. Only a power cycle with unplugging the esp will erase them. Also please post the non truncated log instead of snippets. Changing the mode to off should also close the valve immediately and fully so something does not add up

I made a power cycle before logging, I had to make it in order to connect the ESP to my PC...

**Comment by Anycubic** [comment]
Just noticed this one in the compiler window:

```
src/heating/RadiatorValve.cpp: In member function 'long unsigned int open_heat::heating::RadiatorValve::loop()':
src/heating/RadiatorValve.cpp:25:16: warning: unsigned conversion from 'int' to 'short unsigned int' changes value from '70000' to '4464' [-Woverflow]
   25 |     closeValve(VALVE_FULL_ROTATE_TIME);
      |                ^~~~~~~~~~~~~~~~~~~~~~
src/heating/RadiatorValve.cpp:31:15: warning: unsigned conversion from 'int' to 'short unsigned int' changes value from '70000' to '4464' [-Woverflow]
   31 |     openValve(VALVE_FULL_ROTATE_TIME);
```
You see the same 4464. So `VALVE_FULL_ROTATE_TIME` is always mistakenly converted in 4464?

**Comment by alexmohr** [comment]
If you changed the valuve above 2^16 you will see this. It's changed already on the dev branch

**Comment by Anycubic** [comment]
When did you fix it? I compiled dev branch last night and so the warning

**Comment by alexmohr** [comment]
I just checked and it's not pushed yet. I'll do it later.

**Comment by alexmohr** [comment]
I don't know in which horrible state the dev branch was. It's up to date now. It also now contains the changes to the readme now. Merging via the github ui is ... strange and I should stick to the command line :sweat:

**Comment by Anycubic** [comment]
Yep, I also had problem merging your branch with my test changes, had to re-start from scratch! I'll test it + serial monitor ;-)

**Comment by Anycubic** [comment]
I'm testing last dev which has not the bug "not switching off anymore". It ha also your [5709e47](https://github.com/alexmohr/open-heat/commit/5709e47872ca90db8d92ff0c40167ee7cc260c6c) commit "prevent underflow in valve operation" BUT `currentRotateTime` is still very able to go negative. I saw that happening in the log, when valve is closing after reaching the temperature. Wasn't your commit going to fix this or I didn't understand the fix? I mean if `currentRotateTime` is negative should mean "valve is closed" and put it to "zero" value.

**Comment by Anycubic** [comment]
As you can see here, `currentRotateTime` gets negative and keeps on building up, so when temps goes down and valve is trying to open nothing happen until `currentRotateTime` gets again positive.

![2021-11-16_20-19-07](https://user-images.githubusercontent.com/31375394/142051226-f164530e-51b0-4a48-84f5-223b59df9e38.jpg)

**Comment by Anycubic** [comment]
Or...does actually the valve will open when an open command is issued and `currentRotateTime` is negative? Unfortunately there is no way to see if valve is rotating (well, you can listen to the rumor but you need to have a lot of silence in your room). Maybe a pulsing led when opening and closing could help in debugging situation

**Comment by alexmohr** [comment]
`currentRotateTime` is supposed to be negative and positive. 
The large diff you see is probably from the increased closing times. I'm currently reworking the regulator. Because it's too complicated. 

Let me explain the `currentRotateTime` thingy:
`VALVE_FULL_ROTATE_TIME=60s` for examples sake
if `currentRotateTime` == `-60s`: valve is fully closed.
if `currentRotateTime` == `60s`: valve is fully open.
if `currentRotateTime` == `0s`: valve is exactly in the middle position

Of course these will *not* fit the real positions. But if we rotated for 120s (2x60 from fully open) we are sure it is closed. If we are at `-60` and get a request to close we drop this request as we are already closed. 
If the request is "open" we increment the `currentRotateTime` by the amount of seconds we opened.

**Comment by Anycubic** [comment]
You actually read in my mind, I was going to ask for exactly what you wrote. I don't understand why you did not make it work like 0 <= `currentRotateTime` <= `VALVE_FULL_ROTATE_TIME` (so 0s valve is fully closed etc etc) but I guess there is a very good reason.

BTW even if it is complicated now it seems to me that the regulator is working fine, I switched off the valve after a couple of hours and `abs(currentRotateTime + time needed to close the valve until full closed)` == `VALVE_FULL_ROTATE_TIME`.
I'm working with a "real" `VALVE_FULL_ROTATE_TIME` of 45s and everything seems working good!

**Comment by alexmohr** [comment]
> it seems to me that the regulator is working fine,

you better stop updating then :smile:

**Comment by Anycubic** [comment]
> > it seems to me that the regulator is working fine,
> 
> you better stop updating then 😄

No, no! Better YOU stop changing the code 😄

**Comment by alexmohr** [comment]
Could you post the patch you've currently applied?

**Comment by Anycubic** [comment]
Actually the patch was already committed by you time ago:

```
if (predictDiff <= 2) {
        closeTime = 5000;
      } else if (predictDiff <= 1.5) {
        closeTime = 4000;
      } else if (predictDiff <= 1) {
        closeTime = 2500;
      } else if (predictDiff <= 0.5) {
        closeTime = 1500;
```
I'm also testing these values right now:

```
unsigned short openTime = 300U;
unsigned short closeTime = 400U;
```

**Comment by alexmohr** [comment]
I'm trying to get rid of these two if clauses. 
If it works (and I see that tomorrow if it's cold enough to heat) it will look like this
```
  const auto rotateFactor = 2000.0f;
  const auto maxRotateTime = 10'000;
  auto rotateTime = std::abs(predictDiff) * rotateFactor;
  if (rotateTime > maxRotateTime) {
    rotateTime = maxRotateTime;
  }
.... 
if (predictedTempTooLow) {
    openValve(rotateTime);
  } else {
    // always close a bit more than we open
    closeValve(rotateTime + 500);
  }
```

**Comment by Anycubic** [comment]
Now I'm working with another source of heat: an iron 🤣 tomorrow I'll have a look

**Comment by Anycubic** [comment]
> I'm trying to get rid of these two if clauses. If it works 
> ```

Nice and elegant solution, finding those right const numbers will be crucial. It could be also easier to configure the regulator with different size radiators

**Comment by Anycubic** [comment]
> Maybe instead of `closeValve(rotateTime + 500);` a separate factor would be even better 🤔

What do you mean with separate factor? I told you, I'm a shitty coder

**Comment by alexmohr** [comment]
I'm currently testing new algorithm with a sleep time of 10 minutes which should increase battery life time by a lot. 
I meant the following by separate factor:
```
  const auto rotateFactorOpen = 3000.0f;
  const auto rotateFactorClose = rotateFactorOpen + 200.0f;
```
I'm testing with this at the moment and it works pretty well so far.

**Comment by Anycubic** [comment]
I see, that's better because it is more flexible and taking into account that closing should be executed faster than opening. Just push it and I'll test it as well

**Comment by Anycubic** [comment]
Hi! I have to say I don't like that much new linear regulator algo. Maybe it's nicer and cleaner BUT I have again overshoot and worst thing I tried to change numbers here and there and it is alway overshooting (in a linear way!) :-D
I mean the problem seems the algo which is not very sensible to parameters change

**Comment by Anycubic** [comment]
As you can see temp is always going up in a linear way, I shut off the valve otherwise temp was continuing to go up with the same trend
![2021-11-23_00-39-47](https://user-images.githubusercontent.com/31375394/142951417-390e73a2-e982-4d6a-b02a-912479289d8b.jpg)
:

**Comment by Anycubic** [comment]
Same radiator, old algo, all day long graph:
![Screenshot_2021-11-19-00-33-30-276_io homeassistant companion android](https://user-images.githubusercontent.com/31375394/142953832-ea5d28a6-2e59-4436-9970-fc0dada040f0.jpg)

**Comment by alexmohr** [comment]
With big temperature differences I also see overshooting. I'm currently working this out

**Comment by Anycubic** [comment]
That's with:

```
const auto rotateFactorOpen = 1500.0f;
const auto rotateFactorClose = rotateFactorOpen + 1000.0f;
```

`rotateFactorClose` can be bigger as you like but the algo is too slow in reacting and will always build up error.

![2021-11-23_08-50-59](https://user-images.githubusercontent.com/31375394/142987863-fb5da44c-bcb9-41d7-9545-9d8a18a23b02.jpg)

**Comment by Anycubic** [comment]
I don't know if it is an easy thing to do, but could you please leave the old algo as an option? Like you nicely did with BMP180. I mean it was a finished piece of code and working good, it is a pity just completely discarding it.

**Comment by Anycubic** [comment]
I was thinking that maybe adding a non linear parameter when `set temperature - get temperature <  0.5` could help in slowing down the error building up

**Comment by alexmohr** [comment]
The current version is working fine for my use case. I probably won't support both algorithms, at least not anytime soon. Adding a non linear parameter might work but I can't really test this at the moment. 
You can still patch the old algorithm in.
The new algorithm also checks only every 5 instead of every 2.5 minutes. That's probably the reason why it's too slow for you big radiator. 
You could try to reduce it back to 2.5 and see if it works. If that's the case I make it configurable. 
I reduced the interval in a last effort to get a decent battery runtime.

**Comment by Anycubic** [comment]
> The current version is working fine for my use case. I probably won't support both algorithms, at least not anytime soon. Adding a non linear parameter might work but I can't really test this at the moment. You can still patch the old algorithm in. The new algorithm also checks only every 5 instead of every 2.5 minutes. That's probably the reason why it's too slow for you big radiator. You could try to reduce it back to 2.5 and see if it works. If that's the case I make it configurable. I reduced the interval in a last effort to get a decent battery runtime.

I was already using the old algo with 5 minutes interval, the screenshot that I posted a few posts ago was already with 5 minutes. If you ant to try to add a non linear parameter I will test, it will be a better use of my time than trying to patch the code with the old ago :-D

**Comment by alexmohr** [comment]
Rolling back to the old regulator entirely also would be an option ¯\_(ツ)_/¯

**Comment by Anycubic** [comment]
I didn't dare to say such a thing! It seemed you fall in love with your new regulator algo :-D

**Comment by Anycubic** [comment]
> Rolling back to the old regulator entirely also would be an option ¯_(ツ)_/¯

Indeed I think the old one is far more flexible in configuration and more responsive to parameter change, so I agree here

**Comment by Anycubic** [comment]
Oh come on, don't be so harsh with your Creature! We are not perfect and perfection is not meant for human beings! I found your work very good and perfect for eveyone needs, I also have 3 different types of radiator (little, medium and large) and once I will find nice parameters for all of them you could just implement a drop down configuration so no need to change anything in the code but using web config only. I was also thinking that would be a nice addition to implement  a sort of "wall thermostat command with display" to enable easy temp adjustment/visualisation/furnace on-off without a phone or computer, but this is another story

**Comment by Anycubic** [comment]
Or maybe old regulator with my values is already good for every size of radiators, I didn't have the chance to test it with all of them

**Comment by alexmohr** [comment]
I guess your medium radiator is ~1x0.6m?

For me the new algorithm works better than the old one and it uses the motor less so it's good for battery and valve lifetime
![image](https://user-images.githubusercontent.com/4541068/143047786-5060d354-a7f2-44bb-9a42-b0376b8174bf.png)

> "wall thermostat command with display"
still only works via mqtt and has the issue of the 15 minute delay

I also reduced the `const auto rotateFactorOpen` to `1200`. If that's still not enough for you try with 1000 or 800 and revert the line below back to the original values.

Please try this first. If this does not work I can create a branch to bring back the old implementation, so you can test the parameters with a medium and small regulators. 
If the same parameters work for all 3 I scrap the new implementation.

**Comment by Anycubic** [comment]
Large: 1,80 x 0,75
Medium: 1,80 x 0,55
Small: 0,75 x 0,60

I'm going to test this one with the large:

```
const auto rotateFactorOpen = 1200.0f;
const auto rotateFactorClose = rotateFactorOpen + 1200.0f;
```

Your plan sounds good, keep you posted.

**Comment by Anycubic** [comment]
With:

```
const auto rotateFactorOpen = 1200.0f;
const auto rotateFactorClose = rotateFactorOpen + 1200.0f;
```

Better but still it's pretty lazy! It overshoot but tends not to correct itself.
I'd really appreciate if you could build a branch with old algo, I'm deploying a least 3 valves in production ;-)
![2021-11-23_22-49-45](https://user-images.githubusercontent.com/31375394/143137352-34ce04d3-2770-4f52-b84d-197b35262f74.jpg)

**Comment by Anycubic** [comment]
Same settings as previous post, medium radiator, that's what I call BIG overshoot! There must be something wrong in the code which prevents inverting the slope:
![2021-11-24_08-54-02](https://user-images.githubusercontent.com/31375394/143197162-d7f7681c-35fe-43d2-ae61-3bd8a2150e98.jpg)

**Comment by Anycubic** [comment]
I'd go back to old algorithm, it was a mature enough to start going in production. In the future if you have time you could try to improve the linear algorithm and I could make comparisons in parallel, but right now it is not usable at all.

**Comment by alexmohr** [comment]
https://github.com/alexmohr/open-heat/pull/30

**Comment by Anycubic** [comment]
Very good, I'm going to test it tonight! :-D

**Comment by Anycubic** [comment]
YES! I like it! Always same big radiator, going to test different sizes.

![2021-11-25_20-31-50](https://user-images.githubusercontent.com/31375394/143493247-316ee6a5-d7a6-436b-9231-1828c3ffad95.jpg)

**Comment by alexmohr** [comment]
It ain't working as good for me so I'll maybe support both algorithms. The temperature is always 0.5 to 1 degree too low I guess your home is better insulated because my temperatures drop way faster. 😅

**Comment by Anycubic** [comment]
Well just decrease closetime numbers, it should fix it. I live in a very ancient house, very big walls. But it's isolated, I mean not getting heat like an apartment in a condo

**Comment by Anycubic** [comment]
To be honest I think that no algo will work universally until it's possible to know where the valve piston is

**Comment by Anycubic** [comment]
That's one of my medium radiator (one having the valve at very hight position). As you can see it had a very short overshoot and corrected right away! Old algo regulator is working great! Please do not touch a single bit anymore! 
![2021-11-26_15-45-58](https://user-images.githubusercontent.com/31375394/143598047-dcc9a56f-8f4a-4964-b203-4ba4dbac5900.jpg)

**Comment by Anycubic** [comment]
I applied these changes:

```
auto closeTime = 400U; //to be more reactive and overshoot less
static constexpr int VALVE_FULL_ROTATE_TIME = 45'000; //to stay on the safe side with valves taking a little bit more than 40 seconds to close
```

**Comment by Anycubic** [comment]
Uhmmm, It seems that Karma got me! The medium radiator still doing great since this morning, the big one went bananas two hours ago, temp was slightly increasing like if the valve was dead. Lowered temp by 2 degrees and the valve doing nothing. Shutted off and stil no noise from the valve. Recycled the power and I heard the valve closing eventually.
Logging time! Could you suggest me a good terminal program running under XP? I have a very old laptop.... both Arduino terminal and xterm sucks! Thank you!

**Comment by alexmohr** [comment]
Try terra term but I don't know if this is any better. 
I can imagine what the logs will look like and I don't think they will help and I don't think there is a solution for this except stop detection. 
If you post the logs make sure to post the complete file without truncating anything

**Comment by Anycubic** [comment]
Yes Master! I will do it! ;-)

**Comment by Anycubic** [comment]
> Try terra term but I don't know if this is any better. I can imagine what the logs will look like and I don't think they will 

Installed Tera Term and log to file seems working fine though in terminal windows I can see only gibberish! Numbers are fine, even some words look ok, but the rest are just nonsense symbols. Go figure...

**Comment by alexmohr** [comment]
Sounds like wrong baudrate but then log file wouldn't work either

**Comment by Anycubic** [comment]
Exactly
